lib/charms/karma_k8s/v0/karma.py

```diff
@@ -0,0 +1,269 @@
+
+    def _on_relation_changed(self, _):
+        self.on.alertmanager_config_changed.emit()
+
+    def _on_relation_departed(self, event: RelationDepartedEvent):
+        self._stored.active_relations -= {event.relation.id}
+        self.on.alertmanager_config_changed.emit()
+
+    @property
+    def config_valid(self) -> bool:
+        # karma will fail starting without alertmanager server(s), which would cause pebble to error out.
+
+        # check that there is at least one alertmanager server configured
+        servers = self.get_alertmanager_servers()
+        logger.info("config_valid: servers = %s", servers)
+        return len(servers) > 0
+
+
+class KarmaConsumer(ConsumerBase):
+    """A "consumer" handler to be used by charms that relate to Karma (the 'requires' side of the 'karma' relation).
+    This library offers the interface needed in order to provide Alertmanager URIs and associated information to the
+    Karma application.
+
+    To have your charm provide URIs to Karma, declare the interface's use in your charm's metadata.yaml file:
+
+    ```yaml
+    requires:
+      karma-dashboard:
+        interface: karma_dashboard
+    ```
+
+    A typical example of importing this library might be
+
+    ```python
+    from charms.alertmanager_karma.v0.karma import KarmaConsumer
+    ```
+
+    In your charm's `__init__` method:
+
+    ```python
+    self.karma_lib = KarmaConsumer(
+        self,
+        "karma-dashboard",
+        consumes={"karma": ">=0.0.1"},
+    )
+    ```
+
+    The consumer charm is expected to set config via the consumer library, for example in config-changed:
+
+        if not self.karma_lib.set_config(config):
+            logger.warning("Invalid config: %s", config)
+
+    The consumer charm can then obtain the configured IP address, for example:
+
+        self.unit.status = ActiveStatus("Proxying {}".format(self.karma_lib.ip_address))
+
+    Arguments:
+            charm (CharmBase): consumer charm
+            relation_name (str): from consumer's metadata.yaml
+            consumes (dict): provider specifications
+            multi (bool): multiple relations flag
+
+    Attributes:
+            charm (CharmBase): consumer charm
+    """
+
+    _stored = StoredState()
+
+    def __init__(self, charm, relation_name: str, consumes: dict, multi: bool = False):
+        super().__init__(charm, relation_name, consumes, multi)
+        self.charm = charm
+        self._consumer_relation_name = relation_name  # from consumer's metadata.yaml
+        self._stored.set_default(config={})
+
+        events = self.charm.on[self._consumer_relation_name]
+
+        self.framework.observe(events.relation_joined, self._on_relation_joined)
+
+    def _on_relation_joined(self, event: RelationJoinedEvent):
+        if not self.model.unit.is_leader():
+            return
+
+        # update app data bag
+        event.relation.data[self.charm.unit].update(self._stored.config)
```

Comment:
Since this is a unit data bag, you don't need the guard above

Reply:
Good catch! Used to be app data bag. Updated comment as well.